### PR TITLE
[Enhancement] support element expr prune subfield

### PR DESCRIPTION
**Pull request by Seaven**
Fixes #issue

1. For support rewrite unpack semi-type feature later, refactor `PruneSubfieldRule.java`, split inner class to `SubfieldAccessPathComputer.java`, `SubfieldAccessPathNormalizer.java`, `SubfieldExpressionCollector.java`, and move to dir `prunesubfield`

2.  support prune subfield on the children of ARRAY/MAP's value-column, like: the type of `a2` is `ARRAY<STRUCT<a int, b int>>`
```
MySQL test> explain verbose  select a2[1].a from sc3;
+---------------------------------------------------------------+
| Explain String                                                |
+---------------------------------------------------------------+
| RESOURCE GROUP: default_wg                                    |
|    ......                                                     |
|   |                                                           |
|   0:OlapScanNode                                              |
|      table: sc3, rollup: sc3                                  |
|      preAggregation: on                                       |
|      partitionsRatio=1/1, tabletsRatio=3/3                    |
|      tabletList=7265851,7265853,7265855                       |
|      actualRows=3, avgRowSize=17.0                            |
|      Pruned type: 3 <-> [ARRAY<struct<a int(11), b int(11)>>] |
|      ColumnAccessPath: [/a2/INDEX/a]                          |
|      cardinality: 3                                           |
+---------------------------------------------------------------+
```

3. support more prune mode(INDEX) on ARRAY/MAP, for support choose special index/key on array/map
4. support prune ARRAY's element-column on method `ARRAY_LENGTH`, `CARDINALITY`
5. `ColumnAccessPath` support `is_from_predicate` field, to mark the path is normalized from scan's predicate, for support  late materialize the children of Struct type 
6. refactor `ColumnAccessPath` handle process and the interface of column_iterater on BE, for support  late materialize the children of Struct type

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 270 / 275 (98.18%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/ColumnAccessPath.java|12|13|92.31%|[135]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathNormalizer.java|96|98|97.96%|[123, 163]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathComputer.java|98|100|98.00%|[129, 130]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/ScanNode.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldExpressionCollector.java|22|22|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/PruneSubfieldRule.java|8|8|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 270 / 275 (98.18%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/ColumnAccessPath.java|12|13|92.31%|[135]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathNormalizer.java|96|98|97.96%|[123, 163]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathComputer.java|98|100|98.00%|[129, 130]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/ScanNode.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldExpressionCollector.java|22|22|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/PruneSubfieldRule.java|8|8|100.00%|[]|

File: be/src/column/column_access_path.cpp [hunk 1]
```diff
@@ -84,20 +86,54 @@ StatusOr<ColumnAccessPathPtr> ColumnAccessPath::convert_by_index(const Field* fi
...
+        }
     }
 
     return path;
 }
 
+const std::string& ColumnAccessPath::to_string() const {
+    std::stringstream ss;
+    ss << _path << "(" << _type << ")";
+    return _path;
```
Comment (fzhedu) [inline be/src/column/column_access_path.cpp:136]: ?

File: be/src/column/column_access_path.h [hunk 2]
```diff
@@ -53,9 +53,19 @@ class ColumnAccessPath {
 
     bool is_offset() { return _type == TAccessPathType::type::OFFSET; }
 
+    bool is_field() { return _type == TAccessPathType::type::FIELD; }
+
+    bool is_all() { return _type == TAccessPathType::type::ALL; }
+
+    bool is_index() { return _type == TAccessPathType::type::INDEX; }
```
Comment (fzhedu) [inline be/src/column/column_access_path.h:60]: const

File: be/src/storage/rowset/array_column_iterator.h [hunk 3]
```diff
@@ -56,6 +55,9 @@ class ArrayColumnIterator final : public ColumnIterator {
     std::unique_ptr<ColumnIterator> _null_iterator;
     std::unique_ptr<ColumnIterator> _array_size_iterator;
     std::unique_ptr<ColumnIterator> _element_iterator;
+    ColumnAccessPath* _path;
```
Comment (kangkaisen) [inline be/src/storage/rowset/array_column_iterator.h:58]: what's the path meaning for array column? like these a[1], a[a[4]]?

File: be/src/storage/rowset/segment_iterator.cpp [hunk 4]
```diff
@@ -384,6 +389,7 @@ Status SegmentIterator::_init() {
     // init stage
     // The main task is to do some initialization,
     // initialize the iterator and check if certain optimizations can be applied
+    _init_column_paths();
```
Comment (kangkaisen) [inline be/src/storage/rowset/segment_iterator.cpp:392]: _init_column_access_paths

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/PruneComplexSubfieldTest.java [hunk 5]
```diff
@@ -229,6 +231,70 @@ public void testPruneMapValues() throws Exception {
...
         plan = getVerboseExplain(sql);
-        assertContains(plan, "[/map1/KEY, /map1/OFFSET]");
+        assertContains(plan, "ColumnAccessPath: [/map1/KEY]");
+    }
+
+    @Test
+    public void testPruneComplexFunction() throws Exception {
+        String sql = "select st6.m2[1].s3, st6.a3[2].s6 from sc0";
+        String plan = getVerboseExplain(sql);
+        assertContains(plan, "ColumnAccessPath: [/st6/a3/INDEX/s6, /st6/m2/INDEX/s3]");
```
Comment (kangkaisen) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/PruneComplexSubfieldTest.java:241]: is "[/st6/a3/INDEX/s6" the  a standard representation?

File: be/src/storage/rowset/array_column_iterator.h (hunk 3 above)
Comment (Seaven) [reply be/src/storage/rowset/array_column_iterator.h:58]: yes, like a[1], the path will be `a/INDEX`, I think i can add special paramters in path, to describe the `[]`'s value, but I have not implement now

File: be/src/storage/rowset/column_reader.cpp [hunk 6]
```diff
@@ -466,7 +466,23 @@ StatusOr<std::unique_ptr<ColumnIterator>> ColumnReader::new_iterator(ColumnAcces
         return std::make_unique<ScalarColumnIterator>(this);
     } else if (_column_type == LogicalType::TYPE_ARRAY) {
         size_t col = 0;
-        ASSIGN_OR_RETURN(auto element_iterator, (*_sub_readers)[col++]->new_iterator());
+
+        ColumnAccessPath* value_path = nullptr;
+        if (path != nullptr && !path->children().empty()) {
+            for (const auto& child : path->children()) {
+                if (child->is_index() || child->is_all()) {
+                    if (UNLIKELY(value_path != nullptr)) {
```
[the remark this answers is not in the document]
Comment (Seaven) [reply be/src/storage/rowset/column_reader.cpp:474]: yeah, the children must only one `index` or `all`

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/PruneComplexSubfieldTest.java (hunk 5 above)
Comment (Seaven) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/PruneComplexSubfieldTest.java:241]: None standard representation, just defined by myself

File: be/src/storage/rowset/column_reader.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (Seaven) [reply be/src/storage/rowset/column_reader.cpp:474]: I can modify it

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 265 / 270 (98.15%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/ColumnAccessPath.java|12|13|92.31%|[135]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathNormalizer.java|91|93|97.85%|[105, 149]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathComputer.java|98|100|98.00%|[129, 130]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/ScanNode.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldExpressionCollector.java|22|22|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/PruneSubfieldRule.java|8|8|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 265 / 270 (98.15%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/ColumnAccessPath.java|12|13|92.31%|[135]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathNormalizer.java|91|93|97.85%|[105, 149]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathComputer.java|98|100|98.00%|[129, 130]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/ScanNode.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldExpressionCollector.java|22|22|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/PruneSubfieldRule.java|8|8|100.00%|[]|

File: be/src/storage/rowset/array_column_iterator.cpp [hunk 7]
```diff
@@ -83,7 +82,11 @@ Status ArrayColumnIterator::next_batch(size_t* n, Column* dst) {
     num_to_read = end_offset - num_to_read;
 
     // 3. Read elements
-    RETURN_IF_ERROR(_element_iterator->next_batch(&num_to_read, array_column->elements_column().get()));
+    if (_access_values) {
+        RETURN_IF_ERROR(_element_iterator->next_batch(&num_to_read, array_column->elements_column().get()));
+    } else {
+        array_column->elements_column()->append_default(num_to_read);
```
Comment (fzhedu) [inline be/src/storage/rowset/array_column_iterator.cpp:88]: append_null is also OK?

Comment (Seaven) [reply be/src/storage/rowset/array_column_iterator.cpp:88]: yeah, is same

Comment (Seaven) [reply be/src/storage/rowset/array_column_iterator.cpp:88]: I will remove append-value later, need another PR

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 8]
```diff
@@ -647,6 +650,36 @@ public PlanFragment visitPhysicalDecode(OptExpression optExpression, ExecPlan co
...
+            for (ColumnRefOperator key : scan.getColRefToColumnMetaMap().keySet()) {
+                if (!key.getType().isComplexType()) {
+                    continue;
+                }
+
+                String name = scan.getColRefToColumnMetaMap().get(key).getName();
+                ColumnAccessPath path = normalizer.normalizePath(key, name);
+                if  (path.onlyRoot()) {
+                    continue;
+                }
```
Comment (fzhedu) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:675]: formated

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathNormalizer.java [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+
+        return null;
+    }
+
+    @Override
+    public Void visitIsNullPredicate(IsNullPredicateOperator predicate, Void context) {
+        predicate.getChild(0).accept(this, context);
+        if (currentPath != null) {
+            if (predicate.getChild(0).getType().isMapType()) {
+                currentPath.appendPath(ColumnAccessPath.PATH_PLACEHOLDER, TAccessPathType.OFFSET);
```
Comment (fzhedu) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathNormalizer.java:189]: judge is null according to offset? offset difference is 0 means is null? but that is not true

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathNormalizer.java:189]: judge null by null-column, not offset, set offset just reuse flag, don't need a new flag to mark only need null-column

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 265 / 270 (98.15%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/ColumnAccessPath.java|12|13|92.31%|[135]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathNormalizer.java|91|93|97.85%|[105, 149]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathComputer.java|98|100|98.00%|[129, 130]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/ScanNode.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldExpressionCollector.java|22|22|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/PruneSubfieldRule.java|8|8|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldExpressionCollector.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Override
+    public Void visitSubfield(SubfieldOperator subfieldOperator, Void context) {
+        complexExpressions.add(subfieldOperator);
+        return null;
+    }
+
+    @Override
+    public Void visitCall(CallOperator call, Void context) {
```
Comment (fzhedu) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldExpressionCollector.java:76]: what about lambda expr?

File: fe/fe-core/src/main/java/com/starrocks/planner/ScanNode.java [hunk 11]
```diff
@@ -100,13 +99,22 @@ public String toString() {
...
 
     protected String explainColumnAccessPath(String prefix) {
-        if (CollectionUtils.isEmpty(columnAccessPaths)) {
-            return "";
+        String result = "";
+        if (columnAccessPaths.stream().anyMatch(c -> !c.isFromPredicate())) {
+            result += prefix + "ColumnAccessPath: [" + columnAccessPaths.stream()
+                    .filter(c -> !c.isFromPredicate())
+                    .map(ColumnAccessPath::explain)
+                    .sorted()
```
Comment (fzhedu) [inline fe/fe-core/src/main/java/com/starrocks/planner/ScanNode.java:107]: are they unique?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathComputer.java [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+    @Override
+    public OptExpression visit(OptExpression optExpression, Void context) {
+        visitPredicate(optExpression);
+        return visitChildren(optExpression, context);
+    }
+
+    @Override
+    public OptExpression visitLogicalProject(OptExpression optExpression, Void context) {
+        LogicalProjectOperator lpo = optExpression.getOp().cast();
+        rewriteColumnsInPaths(lpo.getColumnRefMap(), true);
```
Comment (fzhedu) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathComputer.java:75]: is it also applied to common expression in project?

File: be/src/storage/rowset/array_column_iterator.cpp [hunk 13]
```diff
@@ -15,33 +15,32 @@
...
+#include "column/column_access_path.h"
 #include "column/nullable_column.h"
 #include "storage/rowset/scalar_column_iterator.h"
 
 namespace starrocks {
 
 ArrayColumnIterator::ArrayColumnIterator(ColumnReader* reader, std::unique_ptr<ColumnIterator> null_iterator,
                                          std::unique_ptr<ColumnIterator> array_size_iterator,
-                                         std::unique_ptr<ColumnIterator> element_iterator)
+                                         std::unique_ptr<ColumnIterator> element_iterator, ColumnAccessPath* path)
```
Comment (fzhedu) [inline be/src/storage/rowset/array_column_iterator.cpp:26]: will path will be modified? prefer using const

File: fe/fe-core/src/main/java/com/starrocks/planner/ScanNode.java (hunk 11 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/planner/ScanNode.java:107]: yeah

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathComputer.java (hunk 12 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathComputer.java:75]: here has no common expression

File: be/src/storage/rowset/array_column_iterator.cpp (hunk 13 above)
Comment (Seaven) [reply be/src/storage/rowset/array_column_iterator.cpp:26]: it's assgin to class's paramter, const is unnecessary?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldExpressionCollector.java (hunk 10 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldExpressionCollector.java:76]: Lambda expression shouldn't to prune subfield? (I'm not sure, we can support it if lambda expression can be prune subfield) and this doesn't affect prune his input.

File: be/src/storage/rowset/array_column_iterator.cpp (hunk 13 above)
Comment (fzhedu) [reply be/src/storage/rowset/array_column_iterator.cpp:26]: `ColumnAccessPath* const path` to avoid changing the pointer of `path`

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

coverage check fail. please retry. :fearful:

[Please let me know](https://github.com/naver/cover-checker/issues/new) when error again.

com.naver.nid.cover.parser.coverage.exception.ParseException: java.nio.file.NoSuchFileException: /home/runner/_work/starrocks/starrocks/result

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 268 / 270 (99.26%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathNormalizer.java|91|93|97.85%|[105, 149]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/ScanNode.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldExpressionCollector.java|22|22|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ColumnAccessPath.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathComputer.java|100|100|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/prunesubfield/PruneSubfieldRule.java|8|8|100.00%|[]|

**Review by fzhedu — APPROVED** [review]
LGTM

**Review by kangkaisen — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**